### PR TITLE
modify redfish discovery to optionally catalog EMC elements

### PR DESCRIPTION
**Pull request by zyoung51**
- Modify the redfish discovery task to add newly cataloged
  system/chassis to the graph context
- Modify the EMC redfish catalog to iterate through chassis
  or the specified target
- Add condition job for basic logic handling in the graph flow

### CODE REVIEW

File: lib/jobs/condition.js [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+     function ConditionJob(options, context, taskId) {
+        ConditionJob.super_.call(this, logger, options, context, taskId);
+    }
+    util.inherits(ConditionJob, BaseJob);
+
+    /**
+     * @memberOf ConditionJob
+     */
+    ConditionJob.prototype._run = function run() {
+        var when = this.options.when.toLowerCase();
```
Comment (yyscamper) [inline lib/jobs/condition.js]: Needs an assertion for `options.when` or need a default handling routine if `options.when` is not defined.

File: lib/jobs/emc-redfish-catalog.js [hunk 2]
```diff
@@ -42,10 +42,8 @@ function EmcRedfishCatalogJobFactory(
                                    options,
                                    context,
                                    taskId);
-        assert.string(this.context.target);
-        this.nodeId = this.context.target;
-        this.redfish = new RedfishTool();
-        this.redfish.setup(this.nodeId);
+        this.chassis = this.context.chassis ||  [ this.context.target ] || [];
```
Comment (yyscamper) [inline lib/jobs/emc-redfish-catalog.js]: Change the `chassis` to plural will be better to express its meaning.

Comment (yyscamper) [inline lib/jobs/emc-redfish-catalog.js]: Change the `chassis` to plural will be better to express its meaning.

File: lib/task-data/base-tasks/condition.js [hunk 3]
```diff
@@ -0,0 +1,14 @@
+// Copyright 2015, EMC, Inc.
```
Comment (yyscamper) [inline lib/task-data/base-tasks/condition.js]: 2016

File: spec/lib/jobs/condition-job-spec.js [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    it("should run and finish with success", function() {
+        var job = new ConditionJob({when: 'true'}, {}, graphId);
+        job._run();
+        return job._deferred;
+    });
+
+    it("should run and finish with failure", function() {
+        var job = new ConditionJob({when: 'false'}, {}, graphId);
+        job._run();
+        return job._deferred.should.eventually.be.rejectedWith('condition evaluated to false');
```
Comment (yyscamper) [inline spec/lib/jobs/condition-job-spec.js]: The `eventually` should be put together with `reject`, please see this issue for detail: https://github.com/RackHD/RackHD/issues/54 , I tend to think this should be a issue for Chai-as-Promise, I will submit a issue for it. But before it is fixed, we should avoid doing like this.

Comment (yyscamper) [inline spec/lib/jobs/condition-job-spec.js]: The `eventually` should be put together with `reject`, please see this issue for detail: https://github.com/RackHD/RackHD/issues/54 , I tend to think this should be a issue for Chai-as-Promise, I will submit a issue for it. But before it is fixed, we should avoid doing like this.

Comment (yyscamper) [inline spec/lib/jobs/condition-job-spec.js]: The `eventually` should be put together with `reject`, please see this issue for detail: https://github.com/RackHD/RackHD/issues/54 , I tend to think this should be a issue for Chai-as-Promise, I will submit a issue for it. But before it is fixed, we should avoid doing like this.

**Comment by yyscamper** [comment]
I think at least following comment needs be fixed before merge:
(1) `eventually` should not be together with `reject`

**Comment by tannoa2** [comment]
+1

**Comment by zyoung51** [comment]
test this please

File: lib/jobs/condition.js (hunk 1 above)
Comment (benbp) [inline lib/jobs/condition.js]: @yyscamper the assert is in the constructor.

**Comment by benbp** [comment]
👍

**Pull request merged**